### PR TITLE
Use all.sh and its component list in pre-push hook

### PR DESCRIPTION
**Pull request by mpg**
## Description

Currently there is some level of duplication between all.sh and the pre-push hook: when a fast check is added to all.sh we have to remember to add it to the pre-push hook as well.

This commit adds five components that were previously not run:

- `check_recursion`
- `check_changelog`
- `check_test_cases`
- `check_python_files`
- `check_generate_test_code`

It also changes the behaviour in case in-tree cmake was enabled before pushing or otherwise running the pre-push hook: previously the script failed because check_names is not compatible with cmake, now it silently disables cmake (and succeeds unless there is an actual issue).

## Status
**READY**

## Requires Backporting
Yes, as it's a test script.

- [x] 2.16 #3446
- [x] 2.7 #3447 

## Migrations
NO

## Manuel testing

Though the most important case to test is `tests/scripts/all.sh -q -k 'check_*'` with and without errors, as that's what's used in the pre-push hook, I believe the full test matrix is 2x2x2:

- `--quiet` / `--no-quiet`
- `--keep-going` / `--no-keep-going`
- errors or no errors (an error can be triggerred for example by adding trailing whitespace in `library/aes.c`)

I don't think it's really useful at this point to test anything other that the `check_*` components with `-q`.

### CODE REVIEW

File: tests/git-scripts/pre-push.sh [hunk 1]
```diff
@@ -36,14 +36,13 @@ run_test()
...
     fi
 }
 
-run_test ./tests/scripts/check-doxy-blocks.pl
-run_test ./tests/scripts/check-names.sh
-run_test ./tests/scripts/check-generated-files.sh
-run_test ./tests/scripts/check-files.py
-run_test ./tests/scripts/doxygen.sh
+# could direclty invoke "all.sh 'check_*'" but we want shorter output
+for check in $(tests/scripts/all.sh --list-components | grep '^check_'); do
```
Comment (gilles-peskine-arm) [inline tests/git-scripts/pre-push.sh]: So how about `tests/scripts/all.sh -k 'check_*' | ! grep FAILED`?

Comment (mpg) [reply tests/git-scripts/pre-push.sh]: Maybe I'm overthinking this, but what if `all.sh` fails without writing `FAILED` on `stdout`? For example an unexpected error is caught by `set -e` in `all.sh`. In that case the `! grep` would succeed and the non-zero exit status from `all.sh` would be ignore.

Comment (gilles-peskine-arm) [reply tests/git-scripts/pre-push.sh]: Good point. `{ tests/scripts/all.sh -k 'check_*' || echo "FAILED: unexpected failure in all.sh"; } | ! grep FAILED`

Comment (mpg) [reply tests/git-scripts/pre-push.sh]: Sounds good! I'll update the PR with that.

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -270,6 +272,11 @@ msg()
     else
         current_section="$1"
     fi
+
+    if [ $QUIET -eq 1 ]; then
+        return
+    fi
+
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh]: Missing: in quiet mode, silence the call to `dd` to create `seedfile`.

Is anything else missing? I haven't tried a complete run.

File: tests/git-scripts/pre-push.sh [hunk 3]
```diff
@@ -32,18 +32,9 @@ echo "URL is $URL"
...
+    ! "$@"
 }
-
-run_test ./tests/scripts/check-doxy-blocks.pl
-run_test ./tests/scripts/check-names.sh
-run_test ./tests/scripts/check-generated-files.sh
-run_test ./tests/scripts/check-files.py
-run_test ./tests/scripts/doxygen.sh
+{ tests/scripts/all.sh -k 'check_*' || echo "FAILED all.sh"; } 2>&1 |
+    not grep FAILED
```
Comment (gilles-peskine-arm) [inline tests/git-scripts/pre-push.sh]: Why did you define a wrapper function for `!`?

File: tests/scripts/all.sh [hunk 4]
```diff
@@ -186,6 +187,7 @@ Special options:
   --list-components     List components supported on this platform and exit.
 
 General options:
+  -q|--quiet            Only output component names, and errors if any.
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh]: Please add `--no-quiet`.

Comment (gilles-peskine-arm) [inline tests/scripts/all.sh]: Why did you add `--quiet` in this PR, since you aren't using it for the purpose of this PR?

File: tests/scripts/all.sh [hunk 5]
```diff
@@ -1322,8 +1339,8 @@ component_check_python_files () {
 }
 
 component_check_generate_test_code () {
-    msg "uint test: generate_test_code.py"
-    record_status ./tests/scripts/test_generate_test_code.py
+    msg "unit test: generate_test_code.py"
+    record_status ./tests/scripts/test_generate_test_code.py -q 2>/dev/null
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh]: I don't like hiding stderr. I'd strongly prefer modifying `test_generate_test_code.py` to not print anything to stderr if it succeeds.

File: tests/git-scripts/pre-push.sh (hunk 3 above)
Comment (gilles-peskine-arm) [inline tests/git-scripts/pre-push.sh]: If something fails, the output will be something like `FAILED: 1`. It would be more informative to show what failed. For example by grepping or sedding lines surrounded by `^^^^`.

File: tests/git-scripts/pre-push.sh [hunk 6]
```diff
@@ -32,18 +32,9 @@ echo "URL is $URL"
...
+not() {
+    ! "$@"
 }
-
-run_test ./tests/scripts/check-doxy-blocks.pl
-run_test ./tests/scripts/check-names.sh
-run_test ./tests/scripts/check-generated-files.sh
-run_test ./tests/scripts/check-files.py
-run_test ./tests/scripts/doxygen.sh
+{ tests/scripts/all.sh -k 'check_*' || echo "FAILED all.sh"; } 2>&1 |
```
Comment (gilles-peskine-arm) [inline tests/git-scripts/pre-push.sh]: Note that this also runs `check_python_files`, which we don't run on the CI because it is quietly skipped if `pylint3` is not available and our CI doesn't have `pylint3`. Since we don't run `check_python_files` on CI, it tends to get broken; it's currently broken on both `mbed-crypto/development` and `mbedtls/development`.

File: tests/git-scripts/pre-push.sh [hunk 7]
```diff
@@ -32,18 +32,9 @@ echo "URL is $URL"
```
Comment (gilles-peskine-arm) [inline tests/git-scripts/pre-push.sh]: Preexisting: this script runs on what's currently checked out, not on what's being pushed.

File: tests/scripts/all.sh [hunk 8]
```diff
@@ -1349,8 +1366,16 @@ run_component () {
     # The cleanup function will restore it.
     cp -p "$CONFIG_H" "$CONFIG_BAK"
     current_component="$1"
-    "$@"
+    if [ $QUIET -eq 1 ]; then
+        # msg() is silenced, so just print the component name here
+        echo "${current_component#component_}"
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh]: I'd prefer to know when a component has failed. How about changing the output to print `PASS foo` or `FAIL foo` after `component_foo` has finished running?

File: tests/scripts/all.sh (hunk 4 above)
Comment (mpg) [reply tests/scripts/all.sh]: This is just work-in-progress that I pushed mainly so that the code doesn't exist only on my laptop. It's not complete yet, and obviously the intention behind `--quiet` is to use it in the hook.

File: tests/scripts/all.sh (hunk 5 above)
Comment (mpg) [reply tests/scripts/all.sh]: I agree. I could not easily find a way to do that, though.

File: tests/git-scripts/pre-push.sh (hunk 3 above)
Comment (mpg) [reply tests/git-scripts/pre-push.sh]: I assume because `| !` was rejected by my shell when I tried it.

File: tests/scripts/all.sh (hunk 5 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: I had a look at `unittest` and I didn't find a way to make it truly quiet either. But I really don't want to hide errors in non-quiet mode.

I think the best design here would be to have `2>/dev/null` only in `QUIET` mode, and remove `-q`.

File: tests/git-scripts/pre-push.sh (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/git-scripts/pre-push.sh]: Oh right. `!` goes at the beginning of a pipeline:
```
! { … } 2>&1 | grep FAILED
```

File: tests/git-scripts/pre-push.sh (hunk 6 above)
Comment (mpg) [reply tests/git-scripts/pre-push.sh:39]: You fixed that in the meantime, so marking as resolved.

File: tests/git-scripts/pre-push.sh (hunk 7 above)
Comment (mpg) [reply tests/git-scripts/pre-push.sh:32]: Indeed. I'd rather consider it out of scope for this PR and address it independently.

**Comment by mpg** [comment]
@gilles-peskine-arm I've force-pushed a version with a new history, as I don't think the initial fiddling in `pre-push.sh` had much value. I recommend you review the new history directly as the changes are small anyway.

File: tests/git-scripts/pre-push.sh (hunk 1 above)
Comment (mpg) [reply tests/git-scripts/pre-push.sh]: Marking as resolved as the entire question is avoided by the new approach of with `all.sh -q`.

File: tests/scripts/all.sh (hunk 2 above)
Comment (mpg) [reply tests/scripts/all.sh:298]: I've actually silenced it unconditionally as I found even without `-q`, its output was more confusing than helpful.

File: tests/git-scripts/pre-push.sh (hunk 3 above)
Comment (mpg) [reply tests/git-scripts/pre-push.sh]: Marking as resolved as the entire question is avoided by the new approach of with `all.sh -q`.

File: tests/scripts/all.sh (hunk 5 above)
Comment (mpg) [reply tests/scripts/all.sh]: Ok, so what I did is redirect stderr to stdout unconditionally, which is then redirected to `/dev/null` only in quiet mode. (And I removed the `-q`.)

File: tests/git-scripts/pre-push.sh (hunk 3 above)
Comment (mpg) [reply tests/git-scripts/pre-push.sh]: This question is no longer relevant with the new `all.sh -q` approach, so I'm marking the conversation as resolved.

File: tests/scripts/all.sh (hunk 8 above)
Comment (mpg) [reply tests/scripts/all.sh:1933]: Ok, so with `-k` I made sure the existing output from `record_status` is printed to stderr so it's visible even with `-q`. With `--no-keep-going` however we don't really have a chance to print anything, as `set -e` will terminate the shell anyway. We might be able to do something after your "component as subshells" PR.

File: tests/scripts/all.sh (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh:298]: Fully suppressing the output means suppressing error messages as well, but that's unavoidable without adding complications in one form or another, so ok. It's annoying that there's other no simple, portable way to obtain N bytes without the garbage output from `dd` (`dd status=none` is a GNU coreutils extension, `head -c` is widespread but not available on e.g. OpenBSD, and I'd prefer not to artificially restrict the portability of `all.sh`).

File: tests/scripts/all.sh [hunk 9]
```diff
@@ -1886,13 +1925,18 @@ run_component () {
     # Unconditionally create a seedfile that's sufficiently long.
     # Do this before each component, because a previous component may
     # have messed it up or shortened it.
-    dd if=/dev/urandom of=./tests/seedfile bs=64 count=1
+    dd if=/dev/urandom of=./tests/seedfile bs=64 count=1 >/dev/null 2>&1
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1928]: Please don't redirect stderr in non-quiet mode. If an error occurs (e.g. disk full), we need to see the error message, at least in non-quiet mode. You don't need to redirect stdout since `dd` won't print anything there.

File: tests/scripts/all.sh [hunk 10]
```diff
@@ -1886,13 +1925,18 @@ run_component () {
...
-    "$@"
+    if [ $QUIET -eq 1 ]; then
+        # msg() is silenced, so just print the component name here
+        echo "${current_component#component_}"
+    fi
+    redirect_out "$@"
 
     # Restore the build tree to a clean state.
     cleanup
+    current_component=""
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:1939]: Since we aren't planning to use the component name anymore, shouldn't this be `unset current_component`?

**Comment by mpg** [comment]
@gilles-peskine-arm I believe I've addressed all your feedback, except one that you indicated was not a blocker. This is ready for you to review again.
